File: week-5/code/tests/create-pool.ts
import * as anchor from "@coral-xyz/anchor";
import { Program } from "@coral-xyz/anchor";
import { Amm } from "../target/types/amm";
import { assert, expect } from "chai";
import {
  ASSOCIATED_PROGRAM_ID,
  TOKEN_PROGRAM_ID,
} from "@coral-xyz/anchor/dist/cjs/utils/token";
import { mintToken } from "./utils";
import { getAssociatedTokenAddressSync } from "@solana/spl-token";

describe("create-pool", () => {
  const provider = anchor.AnchorProvider.env();
  anchor.setProvider(provider);

  const program = anchor.workspace.Amm as Program<Amm>;

  const payer = anchor.web3.Keypair.generate();
  let id: anchor.web3.PublicKey;
  let fee = 100;

  let ammPda: anchor.web3.PublicKey;

  let mintAKp: anchor.web3.Keypair;
  let mintBKp: anchor.web3.Keypair;

  let poolPda: anchor.web3.PublicKey;
  let poolAuthorityPda: anchor.web3.PublicKey;

  let mintLiquidityPda: anchor.web3.PublicKey;

  let poolAccountA: anchor.web3.PublicKey;
  let poolAccountB: anchor.web3.PublicKey;

  before(async () => {
    await provider.connection.confirmTransaction(
      await provider.connection.requestAirdrop(
        payer.publicKey,
        anchor.web3.LAMPORTS_PER_SOL * 2
      )
    );

    id = anchor.web3.Keypair.generate().publicKey;

    ammPda = anchor.web3.PublicKey.findProgramAddressSync(
      [Buffer.from("amm"), id.toBuffer()],
      program.programId
    )[0];

    const tx = await program.methods
      .createAmm(id, fee)
      .accounts({
        amm: ammPda,
        admin: payer.publicKey,
        systemProgram: anchor.web3.SystemProgram.programId,
      })
      .signers([payer])
      .rpc();

    console.log("Your transaction signature", tx);

    mintAKp = anchor.web3.Keypair.generate();
    mintBKp = anchor.web3.Keypair.generate();

    await mintToken({
      connection: provider.connection,
      payer,
      receiver: payer.publicKey,
      mint: mintAKp,
      decimals: 6,
      amount: 1000,
    });

    await mintToken({
      connection: provider.connection,
      payer,
      receiver: payer.publicKey,
      mint: mintBKp,
      decimals: 6,
      amount: 2000,
    });

    poolPda = anchor.web3.PublicKey.findProgramAddressSync(
      [
        ammPda.toBuffer(),
        mintAKp.publicKey.toBuffer(),
        mintBKp.publicKey.toBuffer(),
      ],
      program.programId
    )[0];

    poolAuthorityPda = anchor.web3.PublicKey.findProgramAddressSync(
      [
        ammPda.toBuffer(),
        mintAKp.publicKey.toBuffer(),
        mintBKp.publicKey.toBuffer(),
        Buffer.from("authority"),
      ],
      program.programId
    )[0];

    mintLiquidityPda = anchor.web3.PublicKey.findProgramAddressSync(
      [
        ammPda.toBuffer(),
        mintAKp.publicKey.toBuffer(),
        mintBKp.publicKey.toBuffer(),
        Buffer.from("mint_liquidity"),
      ],
      program.programId
    )[0];

    console.log("mintLiquidityPda: ", mintLiquidityPda.toBase58());

    poolAccountA = getAssociatedTokenAddressSync(
      mintAKp.publicKey,
      poolAuthorityPda,
      true
    );

    poolAccountB = getAssociatedTokenAddressSync(
      mintBKp.publicKey,
      poolAuthorityPda,
      true
    );
  });

  it("create-pool", async () => {
    try {
      const tx = await program.methods
        .createPool()
        .accounts({
          payer: payer.publicKey,
          pool: poolPda,
          poolAuthority: poolAuthorityPda,
          mintLiquidity: mintLiquidityPda,
          amm: ammPda,
          mintA: mintAKp.publicKey,
          mintB: mintBKp.publicKey,
          poolAccountA: poolAccountA,
          poolAccountB: poolAccountB,

          systemProgram: anchor.web3.SystemProgram.programId,
          tokenProgram: TOKEN_PROGRAM_ID,
          associatedTokenProgram: ASSOCIATED_PROGRAM_ID,
        })
        .signers([payer])
        .rpc();

      console.log("Your transaction signature", tx);
    } catch (error) {
      console.error(error);
    }

    const poolAccount = await program.account.pool.fetch(poolPda);

    assert(poolAccount.amm.toBase58() === ammPda.toBase58(), "Correct AMM");
    assert(
      poolAccount.mintA.toBase58() === mintAKp.publicKey.toBase58(),
      "Correct mint A"
    );
    assert(
      poolAccount.mintB.toBase58() === mintBKp.publicKey.toBase58(),
      "Correct mint B"
    );
  });
});

File: week-5/code/programs/amm/src/instructions/create_pool.rs
use crate::state::{Amm, Pool};
use anchor_lang::prelude::*;
use anchor_spl::token::TokenAccount;
use anchor_spl::{
    associated_token::AssociatedToken,
    token::{Mint, Token},
};

pub fn create_pool(ctx: Context<CreatePool>) -> Result<()> {
    let pool = &mut ctx.accounts.pool;

    pool.amm = ctx.accounts.amm.key();
    pool.mint_a = ctx.accounts.mint_a.key();
    pool.mint_b = ctx.accounts.mint_b.key();

    Ok(())
}

#[derive(Accounts)]
pub struct CreatePool<'info> {
    #[account(
        seeds = [b"amm", amm.id.as_ref()],
        bump
    )]
    pub amm: Box<Account<'info, Amm>>,

    #[account(
            init,
            payer = payer,
            space = 8 + Pool::INIT_SPACE,
            seeds = [
                amm.key().as_ref(),
                mint_a.key().as_ref(),
                mint_b.key().as_ref()
            ],
            bump,
        )]
    pub pool: Box<Account<'info, Pool>>,

    /// CHECK: Read only authority
    #[account(
        seeds = [
            amm.key().as_ref(),
            mint_a.key().as_ref(),
            mint_b.key().as_ref(),
            b"authority"
        ],
        bump,
    )]
    pub pool_authority: AccountInfo<'info>,

    #[account(
            init,
            payer = payer,
            seeds = [
                amm.key().as_ref(),
                mint_a.key().as_ref(),
                mint_b.key().as_ref(),
                b"mint_liquidity"
            ],
            bump,
            mint::decimals = 6,
            mint::authority = pool_authority
        )]
    pub mint_liquidity: Box<Account<'info, Mint>>,

    pub mint_a: Box<Account<'info, Mint>>,

    pub mint_b: Box<Account<'info, Mint>>,

    #[account(
            init,
            payer = payer,
            associated_token::mint = mint_a,
            associated_token::authority = pool_authority,
        )]
    pub pool_account_a: Box<Account<'info, TokenAccount>>,

    #[account(
            init,
            payer = payer,
            associated_token::mint = mint_b,
            associated_token::authority = pool_authority,
        )]
    pub pool_account_b: Box<Account<'info, TokenAccount>>,

    #[account(mut)]
    pub payer: Signer<'info>,

    pub system_program: Program<'info, System>,
    pub token_program: Program<'info, Token>,
    pub associated_token_program: Program<'info, AssociatedToken>,
}
